### PR TITLE
feat(932390): add shell fork bomb detection rule at PL2

### PR DESCRIPTION
**Pull request by fzipi**
## what

- add new rule 932390 (PL2) to detect shell fork bomb patterns
- create regex-assembly file `932390.ra` for the detection pattern
- add 14 regression tests (10 positive, 4 negative)

## why

- fork bombs like `:(){ :|:& };:` were not detected by existing RCE rules
- existing 932 rules rely on known command word lists, but fork bombs work with any function name
- this rule detects the structural pattern `name() { ... | ... & ... }` regardless of the function name used

## refs

- addresses 9Y4-260316 - Fork Bomb Bypass at PL3, reported by @HackingRepo

### CODE REVIEW

**Comment by HackingRepo** [comment]
thank's, I appreciate it

**Comment by HackingRepo** [comment]
But that `f(){ f & f; };` f will evade the rule and not be detected and also a valid fork bomb @fzipi.

**Comment by HackingRepo** [comment]
also `f(){ f() | f(); }; f` and `f(){ f() & f(); }; f` bypass the rule

**Comment by fzipi** [comment]
Updated regex and added tests.

File: regex-assembly/932390.ra [hunk 1]
```diff
@@ -0,0 +1,23 @@
+##! Please refer to the documentation at
+##! https://coreruleset.org/docs/development/regex_assembly/.
+
+##! Detect shell fork bomb patterns.
+##! A fork bomb defines a function whose body recursively calls itself
+##! via a pipe or background operator, causing exponential process
+##! creation. The structural pattern is:
+##!   name() { ... | or & ... }; name
```
Comment (theseion) [inline regex-assembly/932390.ra]:
```suggestion
##!   <name>() { <name call> <| or &> <name call> <optional & }; <name call>
```

File: regex-assembly/932390.ra [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+##!   :(){ :|:& };:         — classic (pipe + background)
+##!   f(){ f & f; };f       — background only
+##!   f(){ f() | f(); };f   — pipe with explicit calls
+##!   f(){ f() & f(); };f   — background with explicit calls
+
+##! Function name: word chars plus : and . (bash builtins)
+##!> define func-name [\w:.]+
+
+##! Match: name() { ... (| or &) ... }
+{{func-name}}\s*\(\s*\)\s*\{[^}]*[|&][^}]*\}
```
Comment (theseion) [inline regex-assembly/932390.ra]:
```suggestion
{{func-name}}\s*\(\s*\)\s*\{[^}]+[|&][^}]+\}
```

File: rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf [hunk 3]
```diff
@@ -1796,6 +1796,47 @@ SecRule REQUEST_HEADERS:User-Agent|REQUEST_HEADERS:Referer "@pmFromFile unix-she
     setvar:'tx.rce_score=+%{tx.critical_anomaly_score}',\
     setvar:'tx.inbound_anomaly_score_pl2=+%{tx.critical_anomaly_score}'"
 
+# [ Shell fork bomb detection ]
+#
+# This rule detects shell fork bomb patterns, where a function is defined
+# whose body recursively calls itself via a pipe and backgrounds the process,
```
Comment (theseion) [inline rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf]:
```suggestion
# whose body recursively calls itself via pipe and / or background process,
```

File: rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf [hunk 4]
```diff
@@ -1796,6 +1796,47 @@ SecRule REQUEST_HEADERS:User-Agent|REQUEST_HEADERS:Referer "@pmFromFile unix-she
...
+# causing exponential process creation that exhausts system resources.
+#
+# Examples:
+#   :(){ :|:& };:
+#   f(){ f|f& };f
+#   test() { test | test & } test
```
Comment (theseion) [inline rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf]:
```suggestion
#   test() { test | test & }; test
```

File: rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf [hunk 5]
```diff
@@ -1796,6 +1796,47 @@ SecRule REQUEST_HEADERS:User-Agent|REQUEST_HEADERS:Referer "@pmFromFile unix-she
...
+#
+# The detection targets the structural pattern:
+#   <name>() { ... | ... & ... }
```
Comment (theseion) [inline rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf]:
```suggestion
#   <name>() { <name call> <| or &> <name call> <optional & }; <name call>
```

File: regex-assembly/932390.ra (hunk 2 above)
Comment (theseion) [inline regex-assembly/932390.ra]: I think this pattern is easy to bypass:
```bash
:(){${}; :|:& };:
```
The closing brace will prevent detection.

Comment (HackingRepo) [reply regex-assembly/932390.ra]: But your ${} already catched by another rule

Comment (fzipi) [reply regex-assembly/932390.ra]: @theseion Is there an alternative?

Comment (theseion) [reply regex-assembly/932390.ra]: Not a simple one. But @HackingRepo is right, it's probably good enough together with other rules.

File: rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf [hunk 6]
```diff
@@ -1796,6 +1796,47 @@ SecRule REQUEST_HEADERS:User-Agent|REQUEST_HEADERS:Referer "@pmFromFile unix-she
...
+# whose body recursively calls itself via pipe and / or background process,
+# causing exponential process creation that exhausts system resources.
+#
+# Examples:
+#   :(){ :|:& };:
+#   f(){ f|f& };f
+#   test() { test | test & }; test
+#
+# The detection targets the structural pattern:
+#   <name>() { <name call> <| or &> <name call> <optional & }; <name call>
```
Comment (theseion) [inline rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf]:
```suggestion
#   <name>() { <name call> <| or &> <name call> <optional &> }; <name call>
```

File: regex-assembly/932390.ra [hunk 7]
```diff
@@ -0,0 +1,23 @@
+##! Please refer to the documentation at
+##! https://coreruleset.org/docs/development/regex_assembly/.
+
+##! Detect shell fork bomb patterns.
+##! A fork bomb defines a function whose body recursively calls itself
+##! via a pipe or background operator, causing exponential process
+##! creation. The structural pattern is:
+##!   <name>() { <name call> <| or &> <name call> <optional & }; <name call>
```
Comment (theseion) [inline regex-assembly/932390.ra]:
```suggestion
##!   <name>() { <name call> <| or &> <name call> <optional &> }; <name call>
```

**Comment by HackingRepo** [comment]
Ok, let's merge @fzipi.
